### PR TITLE
Dismantle Decl a little bit more

### LINKED ISSUE #20407: Regression with source locations in stack traces on Windows
### Zig Version

0.14.0-dev.40+b8d2323b8

### Steps to Reproduce and Observed Behavior

Line numbers in stack traces have regressed recently. I bisected this down to b8d2323b88e68bb5af061dd7e488f51aa0ae8176.

I can only confirm this on Windows, but this is what I'm seeing:

Incorrect stack trace:

```
thread 25548 panic: reached unreachable code
C:\cygwin64\home\kcbanner\kit\zig\build-stage3-release\lib\zig\std\debug.zig:412:14: 0x567d3b8d in assert (game.dll.obj)
    if (!ok) unreachable; // assertion failure
             ^
C:\Users\kcbanner\projects\ion\src\asset.zig:1001:15: 0x5680d414 in release (game.dll.obj)
C:\Users\kcbanner\projects\ion\game\src\modes\RenderTestMode.zig:54:36: 0x5683233d in init (game.dll.obj)
    errdefer gs.asset_cache.release(cube_handle);
                                   ^
C:\Users\kcbanner\projects\ion\game\src\game.zig:80:83: 0x568378be in init (game.dll.obj)
    var quat_editor_mode: imgui.QuatEditorMode = .degrees;
                                                                                  ^
C:\Users\kcbanner\projects\ion\game\src\game.zig:599:45: 0x5683e274 in loaded (game.dll.obj)

                                            ^
C:\Users\kcbanner\projects\ion\game\src\game.zig:424:16: 0x567d1256 in init (game.dll.obj)
    gs.?.loaded(game_memory, reloaded) catch |err| {
               ^
C:\Users\kcbanner\projects\ion\src\main.zig:388:44: 0xb77684 in main (ion.exe.obj)
    game_loaded = game_loaded and game.init(&game_memory);
                                           ^
C:\cygwin64\home\kcbanner\kit\zig\build-stage3-release\lib\zig\std\start.zig:497:75: 0xb7b39a in main (ion.exe.obj)
    return callMainWithArgs(@as(usize, @intCast(c_argc)), @as([*][*:0]u8, @ptrCast(c_argv)), envp);
                                                                          ^
C:\cygwin64\home\kcbanner\kit\zig\build-stage3-release\lib\zig\libc\mingw\crt\crtexe.c:267:0: 0xc27f41 in __tmainCRTStartup (crt2.obj)
    mainret = _tmain (argc, argv, envp);

C:\cygwin64\home\kcbanner\kit\zig\build-stage3-release\lib\zig\libc\mingw\crt\crtexe.c:188:0: 0xc27f9b in mainCRTStartup (crt2.obj)
  ret = __tmainCRTStartup ();

???:?:?: 0x7ff932f97343 in ??? (KERNEL32.DLL)
???:?:?: 0x7ff93441cc90 in ??? (ntdll.dll)
```

Correct stack trace ():

```
c:\Users\kcbanner\projects\ion>zig-debug build && zig-out\bin\ion.exe
thread 26744 panic: reached unreachable code
C:\cygwin64\home\kcbanner\kit\zig\build-stage3-debug\lib\zig\std\debug.zig:412:14: 0x567d3b8d in assert (game.dll.obj)
    if (!ok) unreachable; // assertion failure
             ^
C:\Users\kcbanner\projects\ion\src\asset.zig:454:15: 0x5680d414 in release (game.dll.obj)
        assert(load_ref_count <= handle_ref_count.*);
              ^
C:\Users\kcbanner\projects\ion\game\src\modes\RenderTestMode.zig:54:36: 0x5683233d in init (game.dll.obj)
    errdefer gs.asset_cache.release(cube_handle);
                                   ^
C:\Users\kcbanner\projects\ion\game\src\game.zig:32:83: 0x568378be in init (game.dll.obj)
                return @unionInit(Mode, field.name, try @field(field.type, "init")(gs));
                                                                                  ^
C:\Users\kcbanner\projects\ion\game\src\game.zig:343:45: 0x5683e274 in loaded (game.dll.obj)
            self.active_mode = try Mode.init(.render_test, self);
                                            ^
C:\Users\kcbanner\projects\ion\game\src\game.zig:424:16: 0x567d1256 in init (game.dll.obj)
    gs.?.loaded(game_memory, reloaded) catch |err| {
               ^
C:\Users\kcbanner\projects\ion\src\main.zig:388:44: 0x8d7684 in main (ion.exe.obj)
    game_loaded = game_loaded and game.init(&game_memory);
                                           ^
C:\cygwin64\home\kcbanner\kit\zig\build-stage3-debug\lib\zig\std\start.zig:497:75: 0x8db39a in main (ion.exe.obj)
    return callMainWithArgs(@as(usize, @intCast(c_argc)), @as([*][*:0]u8, @ptrCast(c_argv)), envp);
                                                                          ^
C:\cygwin64\home\kcbanner\kit\zig\build-stage3-debug\lib\zig\libc\mingw\crt\crtexe.c:267:0: 0x987f41 in __tmainCRTStartup (crt2.obj)
    mainret = _tmain (argc, argv, envp);

C:\cygwin64\home\kcbanner\kit\zig\build-stage3-debug\lib\zig\libc\mingw\crt\crtexe.c:188:0: 0x987f9b in mainCRTStartup (crt2.obj)
  ret = __tmainCRTStartup ();

???:?:?: 0x7ff932f97343 in ??? (KERNEL32.DLL)
???:?:?: 0x7ff93441cc90 in ??? (ntdll.dll
```

I don't have a simple reduction yet, as this is only occurring in a large project of mine. 

The interesting thing is that some of the line numbers are out of range of the actual file contents - for example `asset.zig` is only 834 lines long, but the stack trace shows 1001.

Bisect log:

```
git bisect log
git bisect start
# status: waiting for both good and bad commits
# bad: [2b677d1660018434757f9c9efec3c712675e6c47] Sema: fix performance regression
git bisect bad 2b677d1660018434757f9c9efec3c712675e6c47
# status: waiting for good commit(s), bad commit known
# bad: [1ede3af9f6e5923ed1e42cb33e60c0403539c221] std.zig.target: Remove armeb-windows-gnu and aarch64_be-windows-gnu.
git bisect bad 1ede3af9f6e5923ed1e42cb33e60c0403539c221
# status: waiting for good commit(s), bad commit known
# good: [82a934bb912241809ac8029e1fa5843092a7fdf6] Fix futex2_waitv syscall
git bisect good 82a934bb912241809ac8029e1fa5843092a7fdf6
# good: [1fdf13a14825d6e1bd247754e3677cf546ee4020] AstGen: error for redundant `@inComptime()`
git bisect good 1fdf13a14825d6e1bd247754e3677cf546ee4020
# good: [ccd3cc3266762c1fea93cdc0190eaf71718d9e6a] Parse WASM linker flags as base 10, like clang
git bisect good ccd3cc3266762c1fea93cdc0190eaf71718d9e6a
# bad: [5afd774db35e1e12cc49aab5b7f134b741545b80] std: mark Build.args as const
git bisect bad 5afd774db35e1e12cc49aab5b7f134b741545b80
# bad: [edf14777bae56c3a6a155c59f793dc432656c1de] link.Wasm: correctly fetch source location when decl is a type
git bisect bad edf14777bae56c3a6a155c59f793dc432656c1de
# good: [1eaeb4a0a838a783d2060f4e5b3b26b483b26009] Zcu: rework source locations
git bisect good 1eaeb4a0a838a783d2060f4e5b3b26b483b26009
# bad: [f37d0725fa4c6a0b5e4e9ca4d25e2760dc06764d] Sema: fix callconv detection in generic functions
git bisect bad f37d0725fa4c6a0b5e4e9ca4d25e2760dc06764d
# bad: [b8d2323b88e68bb5af061dd7e488f51aa0ae8176] Sema: eliminate Block.src_decl
git bisect bad b8d2323b88e68bb5af061dd7e488f51aa0ae8176
# first bad commit: [b8d2323b88e68bb5af061dd7e488f51aa0ae8176] Sema: eliminate Block.src_decl
```

### Expected Behavior

The line numbers should match the correct version of the stack trace.

### PR DESCRIPTION
**Pull request by mlugg**
This was going to be a bigger branch, but I accidentally regressed stack traces in my previous PR (sorry folks!) and this fixes it. Should be a pretty simple merge. The second commit here is a nop, just a rename I already had done.

The `return 0; // TODO` in `Decl.navSrcLine` might look sus, but rest assured the old logic was screwed, and no contemporary debugger uses the source line in this context as far as I'm aware, so making it screwed in a different way is fine for now.

Line numbers for container-level declarations are no longer relative, because there is no advantage to this. Previously, they were relative, and `Decl` stored the computed absolute number -- this was objectively worse, as we still had the problem of having to update all those absolute numbers, but we were using an extra 4 bytes per `Decl`. With this system, we can track changes in line numbers more directly, avoiding `Decl` as a middle man. The only relative line numbers in ZIR now are in `dbg_stmt` instructions.

This incidentally resolves #20407 -- I'm guessing that a line number was being treated as relative to the wrong thing somewhere. Whatever the issue was, this seems to have solved it!

### CODE REVIEW

**Comment by mlugg** [comment]
@kcbanner, would you be able to confirm that this branch fixes your more complicated repro?

**Comment by kcbanner** [comment]
@mlugg yep, this change resolves the issue! Thanks for the fix!

**Comment by rofrol** [comment]
AnalSubject? Not a good name.

**Comment by mlugg** [comment]
Do you have a better suggestion?

**Comment by rofrol** [comment]
AnalysisSubject?

**Comment by mlugg** [comment]
Too wordy. (I actually meant to call this `AnalUnit`, it'll be renamed again in a future commit.)
There's [nothing wrong](https://github.com/ziglang/zig/blob/8be95af48011f71f1902dff4ffbf5ea95cf1bcf4/src/all_types.hpp#L1341) with having a name that, yes, could be considered to have a funny meaning. It wasn't named for that; it's just a good name.

**Comment by nektro** [comment]
a difference of 4 characters makes it too wordy?

**Comment by mlugg** [comment]
With all due respect, I'm not particularly interested in the opinions of general Zig users on the naming of an internal compiler datastructure which has been discussed and agreed upon elsewhere.

**Comment by drunderscore** [comment]
With all due respect, your defense of "elsewhere" holds no water unless you actually show where and with who this was discussed and agreed upon.

I have no opinion on this personally, but your cavalier attitude makes me question the accuracy and honesty of the response.

**Comment by rohlem** [comment]
> your cavalier attitude makes me question the accuracy and honesty of the response.

From what I remember the abbreviation has been joked about between compiler developers when discussing planned changes on the Discord (channel `#compiler-devel`; so I believe all core members are aware of it.

> Do you have a better suggestion?

I personally think `ana` (-> `AnaSubject`) is just as understandable, one letter shorter even, and not misreadable as something else. I don't personally care much though.

**Comment by drunderscore** [comment]
Thank you for the channel references, I will read through.

I also do not have enough knowledge nor opinion to weigh in (I've certainly seen it used before). I simply found the proposed rationale strangely articulated and thought to express it.

**Comment by RieBi** [comment]
This is hilarious

**Comment by codingjoe** [comment]
Just donated for your next pint at the pub. Thank you for keeping programming fun 🫶

**Comment by sshmendez** [comment]
With all due respect, this discussion is incredible. I would personally prefer `BootySubject`

**Comment by keenans1** [comment]
With all due respect, can we please make the name more work appropriate? How about BootyAnalYsis? Or CaughtWind? ProcessASSistant? Or even ButtProcessor. Lets be professional about this.

**Comment by sshmendez** [comment]
@keenans1 with all due respect, I'm not particularly interested in the opinions of in the opinions of javascript developers on the naming of an internal compiler datastructure which has been discussed and agreed upon elsewhere.

**Comment by keenans1** [comment]
oops. didnt see it was agreed upon elsewhere. i retract.
